### ISSUE TITLE
Investigating performance issues in philox

### ISSUE THREAD

**Issue by bragadeesh** [ROCmSoftwarePlatform/rocRAND#32]
copying @gargrahul 

@jszuppe could you or someone help here? One of the teams is investigating some performance regression issues in rocRAND with changes in compiler (hcc/hip in ROCm). When running (fox example): benchmark_rocrand_kernel --engine philox --dis uniform-float
we are seeing slowdowns. Could you point to the source kernels that are launched with this command? Is it somewhere here:
https://github.com/ROCmSoftwarePlatform/rocRAND/blob/master/library/src/rng/philox4x32_10.hpp
Also explain some top level details about the kernel, and the kernel launch parameters?

It would also help if you could give some tips/ways to simplify the kernel as we track down the perf issues to compiler changes. Thanks

**Comment by ex-rzr** [comment]
Does this happen only to Philox? And what about `benchmark_rocrand_generate --engine philox --dis uniform-float`?

How significant are these slowdowns?

It would be great if you show benchmarks for all engines and all distributions on the compilers (both old and new results):
```
benchmark_rocrand_kernel --engine all --dis all
benchmark_rocrand_generate --engine all --dis all
```

Regarding this particular benchmark.

The kernel is here: https://github.com/ROCmSoftwarePlatform/rocRAND/blob/master/benchmark/benchmark_rocrand_kernel.cpp#L84

And the main code for Philox is here: https://github.com/ROCmSoftwarePlatform/rocRAND/blob/master/library/include/rocrand_philox4x32_10.h

**Comment by gargrahul** [comment]
@ex-rzr There are minor differences in other benchmark results though philox results for benchmark_rocrand_kernel, with HIP PR 535 https://github.com/ROCm-Developer-Tools/HIP/pull/535 are significantly slower than without HIP PR 535 included.

Attached are detailed results. Philox results on Fiji before and after HIP PR 535 are as follows:-

(Before HIP PR 535)
philox:
  uniform-uint:
      Throughput =  123.534 GB/s, Samples =   30.883 GSample/s, AvgTime (1 trial) =    4.047 ms, Time (all) =   80.949 ms, Size = 134217728
  uniform-float:
      Throughput =  130.166 GB/s, Samples =   32.542 GSample/s, AvgTime (1 trial) =    3.841 ms, Time (all) =   76.825 ms, Size = 134217728
  uniform-double:
      Throughput =  119.575 GB/s, Samples =   14.947 GSample/s, AvgTime (1 trial) =    8.363 ms, Time (all) =  167.259 ms, Size = 134217728
  normal-float:
      Throughput =   78.683 GB/s, Samples =   19.671 GSample/s, AvgTime (1 trial) =    6.355 ms, Time (all) =  127.092 ms, Size = 134217728
  normal-double:
      Throughput =   25.275 GB/s, Samples =    3.159 GSample/s, AvgTime (1 trial) =   39.566 ms, Time (all) =  791.311 ms, Size = 134217728
  log-normal-float:
      Throughput =   63.957 GB/s, Samples =   15.989 GSample/s, AvgTime (1 trial) =    7.818 ms, Time (all) =  156.356 ms, Size = 134217728
  log-normal-double:
      Throughput =   17.524 GB/s, Samples =    2.191 GSample/s, AvgTime (1 trial) =   57.064 ms, Time (all) = 1141.270 ms, Size = 134217728
  poisson:
    lambda 10.0
      Throughput =    2.363 GB/s, Samples =    0.591 GSample/s, AvgTime (1 trial) =  211.597 ms, Time (all) = 4231.938 ms, Size = 134217728
  discrete-poisson:
    lambda 10.0
      Throughput =   51.815 GB/s, Samples =   12.954 GSample/s, AvgTime (1 trial) =    9.650 ms, Time (all) =  192.994 ms, Size = 134217728
  discrete-custom:
      Throughput =   52.257 GB/s, Samples =   13.064 GSample/s, AvgTime (1 trial) =    9.568 ms, Time (all) =  191.362 ms, Size = 134217728

(After HIP PR 535)
philox:
  uniform-uint:
      Throughput =   33.059 GB/s, Samples =    8.265 GSample/s, AvgTime (1 trial) =   15.125 ms, Time (all) =  302.491 ms, Size = 134217728
  uniform-float:
      Throughput =   32.981 GB/s, Samples =    8.245 GSample/s, AvgTime (1 trial) =   15.160 ms, Time (all) =  303.205 ms, Size = 134217728
  uniform-double:
      Throughput =   36.494 GB/s, Samples =    4.562 GSample/s, AvgTime (1 trial) =   27.402 ms, Time (all) =  548.040 ms, Size = 134217728
  normal-float:
      Throughput =   33.265 GB/s, Samples =    8.316 GSample/s, AvgTime (1 trial) =   15.031 ms, Time (all) =  300.617 ms, Size = 134217728
  normal-double:
      Throughput =   25.809 GB/s, Samples =    3.226 GSample/s, AvgTime (1 trial) =   38.747 ms, Time (all) =  774.934 ms, Size = 134217728
  log-normal-float:
      Throughput =   33.157 GB/s, Samples =    8.289 GSample/s, AvgTime (1 trial) =   15.080 ms, Time (all) =  301.598 ms, Size = 134217728
  log-normal-double:
      Throughput =   17.797 GB/s, Samples =    2.225 GSample/s, AvgTime (1 trial) =   56.191 ms, Time (all) = 1123.813 ms, Size = 134217728
  poisson:
    lambda 10.0
      Throughput =    0.527 GB/s, Samples =    0.132 GSample/s, AvgTime (1 trial) =  948.613 ms, Time (all) = 18972.261 ms, Size = 134217728
  discrete-poisson:
    lambda 10.0
      Throughput =   29.536 GB/s, Samples =    7.384 GSample/s, AvgTime (1 trial) =   16.928 ms, Time (all) =  338.568 ms, Size = 134217728
  discrete-custom:
      Throughput =   29.901 GB/s, Samples =    7.475 GSample/s, AvgTime (1 trial) =   16.722 ms, Time (all) =  334.433 ms, Size = 134217728

Do you see any possible reason for this big difference? How can I build philox benchmark with a single dist?

[Post_HIP_535_benchmark_rocrand_generate.txt](https://github.com/ROCmSoftwarePlatform/rocRAND/files/2351573/Post_HIP_535_benchmark_rocrand_generate.txt)
[Post_HIP_535_benchmark_rocrand_kernel.txt](https://github.com/ROCmSoftwarePlatform/rocRAND/files/2351574/Post_HIP_535_benchmark_rocrand_kernel.txt)
[Pre_HIP_535_benchmark_rocrand_generate.txt](https://github.com/ROCmSoftwarePlatform/rocRAND/files/2351575/Pre_HIP_535_benchmark_rocrand_generate.txt)
[Pre_HIP_535_benchmark_rocrand_kernel.txt](https://github.com/ROCmSoftwarePlatform/rocRAND/files/2351576/Pre_HIP_535_benchmark_rocrand_kernel.txt)

**Comment by ex-rzr** [comment]
I can't check HIP with and without PR 535 right now, so here are my theoretical thoughts:

1. Only Philox uses vector types (`uint4`)

2. Philox host-API functions (`benchmark_rocrand_generate`) do not suffer this 75% performance degradation.

3. Philox device-API functions call `next()` while host-API functions call `next4()`: https://github.com/ROCmSoftwarePlatform/rocRAND/blob/master/library/include/rocrand_philox4x32_10.h#L196

4. I suspect that the line 198 causes the slowdown: https://github.com/ROCmSoftwarePlatform/rocRAND/blob/master/library/include/rocrand_philox4x32_10.h#L198

There is a dynamic indexing of an array stored in registers. This probably generates `buffer_load` instructions in ISA or something like that.

Can you replace this line with this code and run the benchmarks again?

```cpp
        unsigned int ret;
        switch(m_state.substate)
        {
            case 0:
                ret = m_state.result.x;
                break;
            case 1:
                ret = m_state.result.y;
                break;
            case 2:
                ret = m_state.result.z;
                break;
            default:
                ret = m_state.result.w;
                break;
        }
```

If it works I'll create a PR.

**Comment by gargrahul** [comment]
Thanks @ex-rzr . Your suggestion works well. Please go ahead with PR.

**Comment by ex-rzr** [comment]
Closed by #33
